server/oauth.ts

```diff
@@ -0,0 +1,362 @@
+) => {
+	const OpenIdClient = await ManageMyAccountOpenIdClient();
+
+	// Encode the returnPath, a state token, and the PKCE code verifier into a state cookie
+	const stateToken = crypto.randomBytes(16).toString('base64');
+	const codeVerifier = generators.codeVerifier();
+	const codeChallenge = generators.codeChallenge(codeVerifier);
+	const codeChallengeMethod = 'S256';
+	const state = {
+		returnPath,
+		stateToken,
+		codeVerifier,
+	};
+	const encodedState = Buffer.from(JSON.stringify(state)).toString('base64');
+	res.cookie(OAuthStateCookieName, encodedState, {
+		...oauthCookieOptions,
+		maxAge: ms('10m'),
+	});
+
+	// generate the /authorize endpoint url which we'll redirect the user to
+	const authorizeUrl = OpenIdClient.authorizationUrl({
+		// The sessionToken from authentication to exchange for session cookie
+		sessionToken,
+		// we send the generated stateParam as the state parameter
+		state: stateToken,
+		// any scopes, by default the 'openid' scope is required
+		scope: scopes.join(' '),
+		// the redirect_uri is the callback location we'll redirect to after authentication
+		redirect_uri: redirectUri, // /oauth/callback
+		// PKCE
+		code_challenge: codeChallenge,
+		code_challenge_method: codeChallengeMethod,
+		response_type: 'code',
+		// A max age of 30 minutes means that the user will be prompted to re-authenticate
+		// after 30 minutes of inactivity
+		max_age: Math.floor(ms('30m') / 1000),
+	});
+
+	// redirect the user to the authorize URL
+	return res.redirect(303, authorizeUrl);
+};
+
+export interface IdapiCookie {
+	key: string;
+	value: string;
+	sessionCookie?: boolean;
+}
+
+export interface IdapiCookies {
+	values: IdapiCookie[];
+	expiresAt: string;
+}
+
+export const exchangeAccessTokenForCookies = async (token: string) => {
+	let idapiConfig: IdapiConfig;
+	try {
+		idapiConfig = await getConfig();
+	} catch (e) {
+		throw new Error('Error loading a valid config');
+	}
+	try {
+		const response = await fetch(
+			`https://${idapiConfig.host}/auth/oauth-token?format=cookies&persist=true`,
+			{
+				method: 'POST',
+				body: JSON.stringify({
+					token,
+				}),
+			},
+		);
+		if (!response.ok) {
+			throw new Error(
+				`Error fetching cookies from IDAPI: status ${response.status}.`,
+			);
+		}
+		const json = await response.json();
+		return json.cookies;
+	} catch (error) {
+		throw new Error(error);
+	}
+};
+
+const idapiCookieOptions = (key: string): CookieOptions => ({
+	domain: `.${conf.DOMAIN}`,
+	httpOnly: !['GU_U', 'GU_SO'].includes(key), // unless GU_U/GU_SO cookie, set to true
+	secure: true,
+	sameSite: 'lax',
+	path: '/',
+});
+
+export const setIDAPICookies = (res: Response, cookies: IdapiCookies) => {
+	const { values } = cookies;
+	values.forEach(({ key, value }) => {
+		res.cookie(key, value, {
+			...idapiCookieOptions(key),
+		});
+	});
+};
+
+// Returns either an object with the verified access and ID tokens, or an empty object.
+export const verifyOAuthCookies = async (
```

Comment:
might be worth mentioning that this is doing local verification, especially if you'll later be adding in remote verification too

Reply:
Done!